src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesCcm.Unix.cs

```diff
@@ -0,0 +1,211 @@
+                    }
+
+                    ciphertextBytesWritten += bytesWritten;
+                }
+                else
+                {
+                    byte nonEmptyArray = 123;
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nonEmptyArray, out ciphertextBytesWritten, ref nonEmptyArray, plaintext.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    if (!Interop.Crypto.EvpCipherFinalEx(ctx, ref nonEmptyArray, out int bytesWritten))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    // Should be 0 but let's still validate it
+                    ciphertextBytesWritten += bytesWritten;
+                }
+
+                if (ciphertextBytesWritten != plaintext.Length)
+                {
+                    // this shouldn't happen
+                    throw new CryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpAesCcmGetTag(ctx, ref MemoryMarshal.GetReference(tag), tag.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+            }
+        }
+
+        public void Decrypt(ReadOnlySpan<byte> nonce, ReadOnlySpan<byte> ciphertext, ReadOnlySpan<byte> tag, Span<byte> plaintext, ReadOnlySpan<byte> associatedData)
+        {
+            AesAEAD.CheckArguments(nonce, plaintext, ciphertext, tag);
+            CheckParameters(nonce.Length, tag.Length);
+
+            ref byte nullRef = ref MemoryMarshal.GetReference(Span<byte>.Empty);
+
+            int keyLength = _keyHandle.Key.Length * 8;
+            using (SafeEvpCipherCtxHandle ctx = Interop.Crypto.EvpCipherCreatePartial(
+                    GetCipher(keyLength),
+                    keyLength,
+                    0))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(ctx);
+
+                if (!Interop.Crypto.EvpAesCcmSetNonceLength(ctx, nonce.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpAesCcmSetTag(ctx, ref MemoryMarshal.GetReference(tag), tag.Length))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherSetKeyAndIV(ctx, ref MemoryMarshal.GetReference(_keyHandle.Key), ref MemoryMarshal.GetReference(nonce), 0 /*decrypting*/))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (associatedData.Length != 0)
+                {
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nullRef, out _, ref nullRef, ciphertext.Length))
```

Comment:
Doesn't this say that we're going to read ciphertext.Length bytes from NULL?

Reply:
no, this is "the way" of passing length in openssl (required in case of CCM) - will add comment